### ISSUE TITLE
How to delete Task Hub

### ISSUE THREAD

**Issue by thomas-schreiter** [Azure/azure-functions-durable-extension#448]
The CI/CD pipeline my team uses sets the task hub name to the build id*. This creates a new hub on each deployment so that in-flight orchestrations don't clash with the new code. Does Durable Functions framework offer an api to delete stale hubs?

Currently, I just don't touch the stale queues and tables, but I don't like that approach. I'm thinking about writing a simple script determines what the stale hubs are, and then deletes the queues ("myhub-control-00", "myhub-workitems") and tables ("myhubInstances", "myhubHistory") of these hubs. The script can run in a timer trigger and run once a night.

\* This is combined with swap deployment. Our orchestrations runs for a minutes in most of the cases, so in-flight orchestrations can finish running on the staging slot after the swap.

**Comment by kashimiz** [comment]
Apologies for the delayed response to this. The short answer is that we currently don't have this functionality, but I think that we should. Given that one of our recommended versioning strategies for side-by-side deployment is to generate a new task hub, it makes sense to make cleaning up stale task hubs easy for users.

We do have Durable-specific commands for the Functions CLI in the works (see azure-functions-core-tools/607) but they a) haven't yet made it into a release and b) only cover deleting the history and instance tables, not the queues.

**Comment by kashimiz** [comment]
cc @cgillum

**Comment by thomas-schreiter** [comment]
Thanks, @cgillum. That command works.

Minor feature request: Let the task hub name default to the one in the host.json file if it exists in the current directory.

**Comment by cgillum** [comment]
@thomas-schreiter I'm glad the command worked for you.

Regarding this.

> Minor feature request: Let the task hub name default to the one in the host.json file if it exists in the current directory.

I thought that was the default behavior. @glennamanns does the `delete-task-hub` command use the task hub name in the local host.json if the command-line parameter is not provided?

**Comment by glennamanns** [comment]
@cgillum This is correct. It should do this for all durable commands, including `delete-task-hub`.

@thomas-schreiter If no hub name is provided on the command line, the functions CLI will attempt to find a hub name in host.json. An example v2 host.json:
```
{
    "version": "2.0",
    "extensions": {
        "http": {
            "routePrefix": ""
        },
        "durableTask": {
            "HubName": "TestingHub"
        }
    }
}
```

**Comment by thomas-schreiter** [comment]
Looks like it's a matter of casing. In my host.json, I use `hubName`, starting with lowercase. When running the command without params, it tries to delete the default name:

```
%> head -6 host.json
{
  "version": "2.0",
  "extensions": {
    "durableTask": {
      "hubName": "Hub20181220"
    }

%> func durable delete-task-hub
Task hub 'DurableFunctionsHub' successfully deleted.
```

When changing it to `HubName` (starting with uppercase), the command works as you described.

I'm using lowercase, because Visual Studio complains that the uppercase version is not allowed by the schema: 
![image](https://user-images.githubusercontent.com/3344291/51265956-d4468880-196e-11e9-840d-5173e5af5b19.png)

**Comment by vengi83644** [comment]
> @cgillum This is correct. It should do this for all durable commands, including `delete-task-hub`.
> 
> @thomas-schreiter If no hub name is provided on the command line, the functions CLI will attempt to find a hub name in host.json. An example v2 host.json:
> 
> ```
> {
>     "version": "2.0",
>     "extensions": {
>         "http": {
>             "routePrefix": ""
>         },
>         "durableTask": {
>             "HubName": "TestingHub"
>         }
>     }
> }
> ```

How do we set the hub name in the CLI? I couldn't find any references in the docs.

**Comment by sijucm** [comment]
I just cannot understand how the "func cli" knows which function-app (or storage account) remove the the task hub from. How to pass the name and connection to the function app? I cannot find any useful documentation on this. 
func durable delete-task-hub --task-hub-name UserTest

**Comment by cgillum** [comment]
@vengi83644 all CLI commands have a `--task-hub-name` parameter that you can specify.  You can see it if you run `func durable`. For example:

```
Usage: func durable [context] <action> [-/--options]

Actions:
delete-task-hub     Delete all storage artifacts in the durable task hub
Could not find local host metadata file 'host.json'
    --connection-string-setting (Optional) Name of the setting containing the storage connection string to use.
    --task-hub-name             (Optional) Name of the Durable Task Hub to use.
```

**Comment by sijucm** [comment]
The way Microsoft creates software and also documents it leaves things in a very difficult state. My advice from experience is not to use Microsoft created software like this (durable functions) with non-microsoft-standard options like nodejs or python or Java. You will get in the trap. Most of the features will not work in different combinations not and Microsoft leaves holes for you to fall in and waste your life. Do not trust Microsoft.

The option "    --connection-string-setting (Optional) Name of the setting containing the storage connection string to use.". They just do not document what that means and in which setting and where. I guess it means the host.json, but then you will be checking where that setting should be. Soon, you will realize that rabbit hole works only if it is this or that if not that or those, etc, etc. Microsoft just creates features that traps you.
